File: python/src/elizaos_plugin_moltbook/actions/vote.py
"""
Vote Action

Upvote or downvote posts and comments on Moltbook.

Python port of TypeScript vote action
"""

import re
import logging
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)


def extract_vote_intent(text: str) -> Dict[str, Any]:
    """Extract vote intent from message text"""
    text_lower = text.lower()
    
    # Determine vote type
    is_downvote = any(keyword in text_lower for keyword in ['downvote', 'down vote', 'thumbs down', '👎'])
    vote_type = 'down' if is_downvote else 'up'
    
    # Determine target type
    is_comment = 'comment' in text_lower
    target_type = 'comment' if is_comment else 'post'
    
    # Extract ID (look for common patterns)
    id_match = re.search(r'(?:id[:\s]+)?([a-f0-9]{8,})', text_lower)
    target_id = id_match.group(1) if id_match else None
    
    return {
        'vote_type': vote_type,
        'target_type': target_type,
        'target_id': target_id
    }


async def validate_vote_action(runtime: Any, message: Any, state: Optional[Any] = None) -> bool:
    """Validate if this is a vote action"""
    text = getattr(message.content, 'text', '') or str(message.content)
    text_lower = text.lower()
    
    # Check for vote intent
    has_vote_intent = any(keyword in text_lower for keyword in [
        'upvote', 'downvote', 'vote', 'up vote', 'down vote', 'thumbs up', 'thumbs down', '👍', '👎'
    ])
    
    has_moltbook_mention = any(keyword in text_lower for keyword in [
        'moltbook', 'molty', 'post', 'comment'
    ])
    
    return has_vote_intent and has_moltbook_mention


async def handle_vote_action(
    runtime: Any,
    message: Any,
    state: Optional[Any],
    options: Any,
    callback: Any
) -> Dict[str, Any]:
    """Handle vote action"""
    
    # Get service
    service = runtime.get_service('moltbook')
    if not service:
        error_msg = 'Moltbook service is not available'
        if callback:
            await callback({'text': error_msg, 'error': True})
        return {'success': False, 'error': error_msg}
    
    # Check authentication
    creds = await service.get_credentials()
    if not creds:
        error_msg = 'Not authenticated with Moltbook.'
        if callback:
            await callback({'text': error_msg, 'error': True})
        return {'success': False, 'error': error_msg}
    
    # Extract intent
    text = getattr(message.content, 'text', '') or str(message.content)
    intent = extract_vote_intent(text)
    
    if not intent['target_id']:
        error_msg = 'Please specify the post or comment ID to vote on.'
        if callback:
            await callback({'text': error_msg})
        return {'success': False, 'error': error_msg}
    
    try:
        target_id = intent['target_id']
        vote_type = intent['vote_type']
        target_type = intent['target_type']
        
        # Call appropriate API
        from ..lib.api import vote_post, vote_comment
        
        if target_type == 'comment':
            success = await vote_comment(
                agent_id=str(runtime.agent_id),
                api_key=creds['apiKey'],
                comment_id=target_id,
                vote=vote_type
            )
        else:
            success = await vote_post(
                agent_id=str(runtime.agent_id),
                api_key=creds['apiKey'],
                post_id=target_id,
                vote=vote_type
            )
        
        if success:
            vote_emoji = '👍' if vote_type == 'up' else '👎'
            result_text = f"{vote_emoji} {vote_type.capitalize()}voted {target_type} on Moltbook!"
            if callback:
                await callback({'text': result_text})
            logger.info(f"{vote_type.capitalize()}voted {target_type}: {target_id}")
            return {
                'success': True,
                'text': result_text,
                'data': {
                    'target_id': target_id,
                    'target_type': target_type,
                    'vote_type': vote_type
                }
            }
        else:
            error_msg = f"Failed to {vote_type}vote {target_type}"
            if callback:
                await callback({'text': error_msg, 'error': True})
            return {'success': False, 'error': error_msg}
            
    except Exception as e:
        error_msg = f"Error voting: {str(e)}"
        logger.error(error_msg)
        if callback:
            await callback({'text': error_msg, 'error': True})
        return {'success': False, 'error': error_msg}


# Action definition for plugin registration
VOTE_ACTION = {
    'name': 'MOLTBOOK_VOTE',
    'similes': ['UPVOTE_MOLTBOOK', 'DOWNVOTE_MOLTBOOK', 'MOLTBOOK_UPVOTE', 'MOLTBOOK_DOWNVOTE'],
    'description': 'Upvote or downvote posts and comments on Moltbook.',
    'validate': validate_vote_action,
    'handler': handle_vote_action
}

File: typescript/src/actions/vote.ts
/**
 * Vote Action
 *
 * Upvote or downvote posts and comments on Moltbook.
 */

import type {
  Action,
  ActionResult,
  HandlerCallback,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { PLUGIN_NAME } from "../constants";
import type { MoltbookService } from "../service";

export const voteAction: Action = {
  name: "MOLTBOOK_VOTE",
  similes: [
    "UPVOTE_MOLTBOOK",
    "DOWNVOTE_MOLTBOOK",
    "MOLTBOOK_UPVOTE",
    "MOLTBOOK_DOWNVOTE",
    "LIKE_MOLTBOOK",
  ],
  description: "Vote on a Moltbook post or comment (upvote, downvote, or remove vote).",

  validate: async (runtime: any, message: any, state?: any, options?: any): Promise<boolean> => {
    const __avTextRaw = typeof message?.content?.text === "string" ? message.content.text : "";
    const __avText = __avTextRaw.toLowerCase();
    const __avKeywords = ["moltbook", "vote"];
    const __avKeywordOk =
      __avKeywords.length > 0 && __avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
    const __avRegex = /\b(?:moltbook|vote)\b/i;
    const __avRegexOk = __avRegex.test(__avText);
    const __avSource = String(message?.content?.source ?? message?.source ?? "");
    const __avExpectedSource = "";
    const __avSourceOk = __avExpectedSource
      ? __avSource === __avExpectedSource
      : Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
    const __avOptions = options && typeof options === "object" ? options : {};
    const __avInputOk =
      __avText.trim().length > 0 ||
      Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
      Boolean(message?.content && typeof message.content === "object");

    if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
      return false;
    }

    const __avLegacyValidate = async (
      _runtime: IAgentRuntime,
      message: Memory,
      _state: State | undefined
    ): Promise<boolean> => {
      const text = message.content.text?.toLowerCase() || "";

      // Check for voting intent
      const hasVoteIntent =
        text.includes("upvote") ||
        text.includes("downvote") ||
        text.includes("vote") ||
        text.includes("like");

      const hasMoltbookMention =
        text.includes("moltbook") || text.includes("post") || text.includes("comment");

      return hasVoteIntent && hasMoltbookMention;
    };
    try {
      return Boolean(await (__avLegacyValidate as any)(runtime, message, state, options));
    } catch {
      return false;
    }
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    _options: unknown,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = runtime.getService<MoltbookService>(PLUGIN_NAME);
    if (!service) {
      const error = "Moltbook service is not available";
      if (callback) {
        await callback({ text: error, error: true });
      }
      return { success: false, error: new Error(error) };
    }

    // Check authentication
    const creds = await service.getCredentials();
    if (!creds) {
      const error = "Not authenticated with Moltbook.";
      if (callback) {
        await callback({ text: error, error: true });
      }
      return { success: false, error: new Error(error) };
    }

    // Check if account is claimed (required to vote)
    if (creds.claimStatus !== "claimed") {
      const claimUrl = creds.claimUrl || "https://moltbook.com";
      const error = `Cannot vote - account not yet claimed by human. Claim URL: ${claimUrl}`;
      if (callback) {
        await callback({
          text: `I can't vote on Moltbook yet - my account needs to be claimed by a human first. Please visit: ${claimUrl}`,
          error: true,
        });
      }
      runtime.logger.warn({ claimUrl }, "Moltbook: Attempted to vote but account not claimed");
      return { success: false, error: new Error(error) };
    }

    // Extract intent
    const intent = extractVoteIntent(message.content.text || "");

    if (!intent.postId) {
      const error = "Please specify which post to vote on (by ID).";
      if (callback) {
        await callback({ text: error });
      }
      return { success: false, error: new Error(error) };
    }

    try {
      let success: boolean;
      let description: string;

      if (intent.commentId) {
        // Vote on comment (Moltbook API uses only commentId)
        success = await service.voteComment(intent.commentId, intent.direction);
        description = `${intent.direction === "up" ? "Upvoted" : "Downvoted"} comment`;
      } else {
        // Vote on post
        success = await service.votePost(intent.postId!, intent.direction);
        description = `${intent.direction === "up" ? "Upvoted" : "Downvoted"} post`;
      }

      if (!success) {
        const error = "Failed to register vote.";
        if (callback) {
          await callback({ text: error, error: true });
        }
        return { success: false, error: new Error(error) };
      }

      // Success response
      if (callback) {
        const emoji = intent.direction === "up" ? "👍" : "👎";
        await callback({
          text: `${emoji} ${description}`,
        });
      }

      return {
        success: true,
        text: description,
        values: {
          postId: intent.postId,
          commentId: intent.commentId,
          direction: intent.direction,
        },
        data: {
          action: "MOLTBOOK_VOTE",
          postId: intent.postId,
          commentId: intent.commentId,
          direction: intent.direction,
        },
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      runtime.logger.error({ error }, "Error voting on Moltbook");

      if (callback) {
        await callback({
          text: `Failed to vote: ${errorMessage}`,
          error: true,
        });
      }

      return {
        success: false,
        error: error instanceof Error ? error : new Error(errorMessage),
      };
    }
  },

  examples: [
    [
      {
        name: "{{userName}}",
        content: {
          text: "Upvote post abc123 on Moltbook",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "👍 Upvoted post",
          actions: ["MOLTBOOK_VOTE"],
        },
      },
    ],
    [
      {
        name: "{{userName}}",
        content: {
          text: "Downvote that Moltbook comment xyz789",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "👎 Downvoted comment",
          actions: ["MOLTBOOK_VOTE"],
        },
      },
    ],
  ],
};

/**
 * Extract vote intent from user message
 * Note: Moltbook API only supports upvote/downvote, not vote removal
 */
function extractVoteIntent(text: string): {
  postId?: string;
  commentId?: string;
  direction: "up" | "down";
} {
  const lowerText = text.toLowerCase();

  // Determine direction (default to upvote)
  const direction: "up" | "down" =
    lowerText.includes("downvote") || lowerText.includes("dislike") ? "down" : "up";

  // Extract post ID
  const postIdMatch = text.match(/post\s+([a-zA-Z0-9_-]+)/i);
  const postId = postIdMatch ? postIdMatch[1] : undefined;

  // Extract comment ID
  const commentIdMatch = text.match(/comment\s+([a-zA-Z0-9_-]+)/i);
  const commentId = commentIdMatch ? commentIdMatch[1] : undefined;

  return { postId, commentId, direction };
}

export default voteAction;
